spec/controllers/admin_request_controller_spec.rb

```diff
@@ -183,7 +183,10 @@
     end
 
     it 'expires the request cache when saving edits to it' do
-      allow(InfoRequest).to receive(:find).with(info_request.id.to_s).and_return(info_request)
+      allow(InfoRequest).to receive(:find).
+        with(info_request.id).and_return(info_request)
+      allow(InfoRequest).to receive(:find).
+        with(info_request.id.to_s).and_return(info_request)
```

Comment:
Annoyingly, It does both **but** further investigation reveals that this is the new Rails 5 behaviour around running tests in transactions (I miss the gem version - I haven't figured out how to selectively disable this yet, assuming that's an option) which means it's running [the commit hook for the related `info_request_event`](https://github.com/mysociety/alaveteli/blob/1f6f56b3d23435830be135f173d35d8e40619609/app/models/info_request_event.rb#L89-L90), resulting in unexpected additional calls to `InfoRequest.find` (with an Integer) as it tries to [build the request summary](https://github.com/mysociety/alaveteli/blob/1f6f56b3d23435830be135f173d35d8e40619609/app/models/alaveteli_pro/request_summary.rb#L53-L66). (The original `allow` is for [the controller `before_action`](https://github.com/mysociety/alaveteli/tree/1f6f56b3d23435830be135f173d35d8e40619609/app/controllers/admin_request_controller.rb#L214-L216) that takes `params[:id]` which is still a String.)

I'm half tempted to withdraw this and bundle it with other test transaction related things.

Alternatively, the better - but still lazy - solution is to stub the call to `InfoRequest#create_or_update_request_summary` instead as we don't actually want it to run here.

```diff
-     allow(InfoRequest).to receive(:find).
-       with(info_request.id).and_return(info_request)
+     allow_any_instance_of(InfoRequest).
+       to receive(:create_or_update_request_summary)
      allow(InfoRequest).to receive(:find).
        with(info_request.id.to_s).and_return(info_request)
```

Reply:
Or ... the more generic technique of allowing the generic request then the specific one

```diff
-     allow(InfoRequest).to receive(:find).
-       with(info_request.id).and_return(info_request)
+     allow(InfoRequest).to receive(:find).and_call_original
      allow(InfoRequest).to receive(:find).
        with(info_request.id.to_s).and_return(info_request)
```

(Turning transactional tests off doesn't seem to make a difference, `after_commit` seems to run in any case)

Reply:
Okay to summarise, as I understand it `and_call_original` has been added to tests because the `after_commit` callback in `InfoRequest` also it calls `InfoRequest.find` but with an integer - where as the controller is calling it with a string from the parameters. And as Rails 5 now always runs `after_commit` callbacks its causing this issue.

Personally I don't like needing to use `and_call_original`. I think it adds complexly to the test and requires knowledge of the production code to really understand what the are doing tests. There are some alternatives we could look at:

1. As suggested, stub `InfoRequest#create_or_update_request_summary` but again not ideal to stub an seemingly unrelated part of the code.
2. Ensure we always find with an integer in the controller:
```diff
diff --git a/app/controllers/admin_request_controller.rb b/app/controllers/admin_request_controller.rb
index 96c83d555..196cff631 100644
--- a/app/controllers/admin_request_controller.rb
+++ b/app/controllers/admin_request_controller.rb
@@ -212,7 +212,7 @@ class AdminRequestController < AdminController
   end
 
   def set_info_request
-    @info_request = InfoRequest.find(params[:id])
+    @info_request = InfoRequest.find(params[:id].to_i)
   end
 
 end
diff --git a/spec/controllers/admin_request_controller_spec.rb b/spec/controllers/admin_request_controller_spec.rb
index add9e3bcb..8fdcecaa8 100644
--- a/spec/controllers/admin_request_controller_spec.rb
+++ b/spec/controllers/admin_request_controller_spec.rb
@@ -183,9 +183,8 @@ describe AdminRequestController, "when administering requests" do
     end
 
     it 'expires the request cache when saving edits to it' do
-      allow(InfoRequest).to receive(:find).and_call_original
       allow(InfoRequest).to receive(:find).
-        with(info_request.id.to_s).and_return(info_request)
+        with(info_request.id).and_return(info_request)
       expect(info_request).to receive(:expire)
       post :update, params: {
                       :id => info_request.id,
@@ -246,9 +245,8 @@ describe AdminRequestController, "when administering requests" do
     end
 
     it 'expires the file cache for the request' do
-      allow(InfoRequest).to receive(:find).and_call_original
       allow(InfoRequest).to receive(:find).
-        with(info_request.id.to_s).and_return(info_request)
+        with(info_request.id).and_return(info_request)
       expect(info_request).to receive(:expire)
       post :hide, params: {
                     :id => info_request.id,

```

3: Use `find_by!` instead of `find`:
```diff
diff --git a/app/controllers/admin_request_controller.rb b/app/controllers/admin_request_controller.rb
index 96c83d555..12be39042 100644
--- a/app/controllers/admin_request_controller.rb
+++ b/app/controllers/admin_request_controller.rb
@@ -212,7 +212,7 @@ class AdminRequestController < AdminController
   end
 
   def set_info_request
-    @info_request = InfoRequest.find(params[:id])
+    @info_request = InfoRequest.find_by!(id: params[:id])
   end
 
 end
diff --git a/spec/controllers/admin_request_controller_spec.rb b/spec/controllers/admin_request_controller_spec.rb
index add9e3bcb..752bdb781 100644
--- a/spec/controllers/admin_request_controller_spec.rb
+++ b/spec/controllers/admin_request_controller_spec.rb
@@ -183,9 +183,8 @@ describe AdminRequestController, "when administering requests" do
     end
 
     it 'expires the request cache when saving edits to it' do
-      allow(InfoRequest).to receive(:find).and_call_original
-      allow(InfoRequest).to receive(:find).
-        with(info_request.id.to_s).and_return(info_request)
+      allow(InfoRequest).to receive(:find_by!).
+        with(id: info_request.id.to_s).and_return(info_request)
       expect(info_request).to receive(:expire)
       post :update, params: {
                       :id => info_request.id,
@@ -246,9 +245,8 @@ describe AdminRequestController, "when administering requests" do
     end
 
     it 'expires the file cache for the request' do
-      allow(InfoRequest).to receive(:find).and_call_original
-      allow(InfoRequest).to receive(:find).
-        with(info_request.id.to_s).and_return(info_request)
+      allow(InfoRequest).to receive(:find_by!).
+        with(id: info_request.id.to_s).and_return(info_request)
       expect(info_request).to receive(:expire)
       post :hide, params: {
                     :id => info_request.id,

```